File: src/java/org/apache/fop/hyphenation/Hyphenator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.hyphenation;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.net.URISyntaxException;
import java.util.Map;

import org.xml.sax.InputSource;

import org.apache.commons.io.IOUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.apache.fop.apps.io.InternalResourceResolver;

/**
 * <p>This class is the main entry point to the hyphenation package.
 * You can use only the static methods or create an instance.</p>
 *
 * <p>This work was authored by Carlos Villegas (cav@uniscope.co.jp).</p>
 */
public final class Hyphenator {

    /** logging instance */
    protected static final Log log = LogFactory.getLog(Hyphenator.class);

    private static HyphenationTreeCache hTreeCache;

    /** Enables a dump of statistics. Note: If activated content is sent to System.out! */
    private static boolean statisticsDump;

    /**
     * Creates a new hyphenator.
     */
    private Hyphenator() { }

    /** @return the default (static) hyphenation tree cache */
    public static synchronized HyphenationTreeCache getHyphenationTreeCache() {
        if (hTreeCache == null) {
            hTreeCache = new HyphenationTreeCache();
        }
        return hTreeCache;
    }

    /**
     * Returns a hyphenation tree for a given language and country,
     * with fallback from (lang,country) to (lang).
     * The hyphenation trees are cached.
     * @param lang the language
     * @param country the country (may be null or "none")
     * @param resolver resolver to find the hyphenation files
     * @param hyphPatNames the map with user-configured hyphenation pattern file names
     * @return the hyphenation tree
     */
    public static HyphenationTree getHyphenationTree(String lang,
            String country, InternalResourceResolver resolver, Map hyphPatNames) {
        String llccKey = HyphenationTreeCache.constructLlccKey(lang, country);
        HyphenationTreeCache cache = getHyphenationTreeCache();

        // If this hyphenation tree has been registered as missing, return immediately
        if (cache.isMissing(llccKey)) {
            return null;
        }

        HyphenationTree hTree = getHyphenationTree2(lang, country, resolver, hyphPatNames);

        // fallback to lang only
        if (hTree == null && country != null && !country.equals("none")) {
            String llKey = HyphenationTreeCache.constructLlccKey(lang, null);
            if (!cache.isMissing(llKey)) {
                hTree = getHyphenationTree2(lang, null, resolver, hyphPatNames);
                if (hTree != null && log.isDebugEnabled()) {
                    log.debug("Couldn't find hyphenation pattern "
                              + "for lang=\"" + lang + "\",country=\"" + country + "\"."
                              + " Using general language pattern "
                              + "for lang=\"" + lang + "\" instead.");
                }
                if (hTree == null) {
                    // no fallback; register as missing
                    cache.noteMissing(llKey);
                } else {
                    // also register for (lang,country)
                    cache.cache(llccKey, hTree);
                }
            }
        }

        if (hTree == null) {
            // (lang,country) and (lang) tried; register as missing
            cache.noteMissing(llccKey);
            log.error("Couldn't find hyphenation pattern "
                      + "for lang=\"" + lang + "\""
                      + (country != null && !country.equals("none")
                              ? ",country=\"" + country + "\""
                              : "")
                      + ".");
        }

        return hTree;
    }

    /**
     * Returns a hyphenation tree for a given language and country
     * The hyphenation trees are cached.
     * @param lang the language
     * @param country the country (may be null or "none")
     * @param resourceResolver resolver to find the hyphenation files
     * @param hyphPatNames the map with user-configured hyphenation pattern file names
     * @return the hyphenation tree
     */
    public static HyphenationTree getHyphenationTree2(String lang,
            String country, InternalResourceResolver resourceResolver, Map hyphPatNames) {
        String llccKey = HyphenationTreeCache.constructLlccKey(lang, country);
        HyphenationTreeCache cache = getHyphenationTreeCache();

        HyphenationTree hTree;
        // first try to find it in the cache
        hTree = getHyphenationTreeCache().getHyphenationTree(lang, country);
        if (hTree != null) {
            return hTree;
        }

        String key = HyphenationTreeCache.constructUserKey(lang, country, hyphPatNames);
        if (key == null) {
            key = llccKey;
        }

        if (resourceResolver != null) {
            hTree = getUserHyphenationTree(key, resourceResolver);
        }
        if (hTree == null) {
            hTree = getFopHyphenationTree(key);
        }

        // put it into the pattern cache
        if (hTree != null) {
            cache.cache(llccKey, hTree);
        }

        return hTree;
    }

    private static InputStream getResourceStream(String key) {
        InputStream is = null;
        // Try to use Context Class Loader to load the properties file.
        try {
            java.lang.reflect.Method getCCL = Thread.class.getMethod(
                    "getContextClassLoader", new Class[0]);
            if (getCCL != null) {
                ClassLoader contextClassLoader = (ClassLoader)getCCL.invoke(
                        Thread.currentThread(),
                        new Object[0]);
                is = contextClassLoader.getResourceAsStream("hyph/" + key
                                                            + ".hyp");
            }
        } catch (NoSuchMethodException e) {
            //ignore, fallback further down
        } catch (IllegalAccessException e) {
            //ignore, fallback further down
        } catch (java.lang.reflect.InvocationTargetException e) {
            //ignore, fallback further down
        }

        if (is == null) {
            is = Hyphenator.class.getResourceAsStream("/hyph/" + key
                                                      + ".hyp");
        }

        return is;
    }

    private static HyphenationTree readHyphenationTree(InputStream in) {
        HyphenationTree hTree = null;
        try {
            ObjectInputStream ois = new ObjectInputStream(in);
            hTree = (HyphenationTree)ois.readObject();
        } catch (IOException ioe) {
            log.error("I/O error while loading precompiled hyphenation pattern file", ioe);
        } catch (ClassNotFoundException cnfe) {
            log.error("Error while reading hyphenation object from file", cnfe);
        }
        return hTree;
    }

    /**
     * Returns a hyphenation tree. This method looks in the resources (getResourceStream) for
     * the hyphenation patterns.
     * @param key the language/country key
     * @return the hyphenation tree or null if it wasn't found in the resources
     */
    public static HyphenationTree getFopHyphenationTree(String key) {
        HyphenationTree hTree = null;
        ObjectInputStream ois = null;
        InputStream is = null;
        try {
            is = getResourceStream(key);
            if (is == null) {
                if (log.isDebugEnabled()) {
                    log.debug("Couldn't find precompiled hyphenation pattern "
                              + key + " in resources");
                }
                return null;
            }
            hTree = readHyphenationTree(is);
        } finally {
            IOUtils.closeQuietly(ois);
        }
        return hTree;
    }

    /**
     * Load tree from serialized file or xml file
     * using configuration settings
     * @param key language key for the requested hyphenation file
     * @param resourceResolver resource resolver to find the hyphenation files
     * @return the requested HypenationTree or null if it is not available
     */
    public static HyphenationTree getUserHyphenationTree(String key,
            InternalResourceResolver resourceResolver) {
        HyphenationTree hTree = null;
        // I use here the following convention. The file name specified in
        // the configuration is taken as the base name. First we try
        // name + ".hyp" assuming a serialized HyphenationTree. If that fails
        // we try name + ".xml", assumming a raw hyphenation pattern file.

        // first try serialized object
        String name = key + ".hyp";
        try {
            InputStream in = getHyphenationTreeStream(name, resourceResolver);
            try {
                hTree = readHyphenationTree(in);
            } finally {
                IOUtils.closeQuietly(in);
            }
            return hTree;
        } catch (IOException ioe) {
            if (log.isDebugEnabled()) {
                log.debug("I/O problem while trying to load " + name, ioe);
            }
        }

        // try the raw XML file
        name = key + ".xml";
        hTree = new HyphenationTree();
        try {
            InputStream in = getHyphenationTreeStream(name, resourceResolver);
            try {
                InputSource src = new InputSource(in);
                src.setSystemId(name);
                hTree.loadPatterns(src);
            } finally {
                IOUtils.closeQuietly(in);
            }
            if (statisticsDump) {
                System.out.println("Stats: ");
                hTree.printStats();
            }
            return hTree;
        } catch (HyphenationException ex) {
            log.error("Can't load user patterns from XML file " + name + ": " + ex.getMessage());
            return null;
        } catch (IOException ioe) {
            if (log.isDebugEnabled()) {
                log.debug("I/O problem while trying to load " + name, ioe);
            }
            return null;
        }
    }

    private static InputStream getHyphenationTreeStream(String name,
            InternalResourceResolver resourceResolver) throws IOException {
        try {
            return new BufferedInputStream(resourceResolver.getResource(name));
        } catch (URISyntaxException use) {
            log.debug("An exception was thrown while attempting to load " + name, use);
        }
        return null;
    }

    /**
     * Hyphenates a word.
     * @param lang the language
     * @param country the optional country code (may be null or "none")
     * @param resourceResolver resolver to find the hyphenation files
     * @param hyphPatNames the map with user-configured hyphenation pattern file names
     * @param word the word to hyphenate
     * @param leftMin the minimum number of characters before the hyphenation point
     * @param rightMin the minimum number of characters after the hyphenation point
     * @return the hyphenation result
     */
    public static Hyphenation hyphenate(String lang, String country,
            InternalResourceResolver resourceResolver, Map hyphPatNames, String word, int leftMin,
            int rightMin) {
        HyphenationTree hTree = getHyphenationTree(lang, country, resourceResolver, hyphPatNames);
        if (hTree == null) {
            return null;
        }
        return hTree.hyphenate(word, leftMin, rightMin);
    }
}

File: dotnet/src/Fop.Hyphenation/Hyphenator.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Reflection;
using System.Xml;

namespace Fop.Hyphenation;

/// <summary>
/// The main entry point to the hyphenation package: resolves and caches a
/// <see cref="HyphenationTree"/> for a language (and optional country) and hyphenates words.
/// <para>
/// Modern reformulation of <c>org.apache.fop.hyphenation.Hyphenator</c>. Where FOP loads precompiled
/// <c>.hyp</c> serialized trees or user XML through a resource resolver, this port loads the XML
/// pattern files bundled as embedded resources (<c>Patterns/{lang}.xml</c>, in the FOP
/// <c>hyphenation.dtd</c> format). Additional languages are added simply by dropping another embedded
/// <c>Patterns/{lang}.xml</c> in. An unknown language resolves to <c>null</c> (no hyphenation), so the
/// caller degrades gracefully.
/// </para>
/// </summary>
public sealed class Hyphenator
{
    private const string ResourcePrefix = "Fop.Hyphenation.Patterns.";

    /// <summary>A shared default instance backed by the bundled embedded patterns.</summary>
    public static Hyphenator Default { get; } = new();

    private readonly Lock gate = new();

    // A null cached value records a language whose patterns are absent, so we don't retry the lookup.
    private readonly Dictionary<string, HyphenationTree?> cache = new(StringComparer.Ordinal);

    /// <summary>
    /// Returns the cached hyphenation tree for <paramref name="lang"/> (and optional
    /// <paramref name="country"/>), loading it from the embedded patterns on first use, or <c>null</c>
    /// when no patterns are bundled for the language. Lookup tries the full <c>lang_country</c> key
    /// first, then falls back to the bare language.
    /// </summary>
    /// <param name="lang">the language code (e.g. "en"), case-insensitive.</param>
    /// <param name="country">an optional country code (e.g. "US"), or <c>null</c>/"none".</param>
    public HyphenationTree? GetHyphenationTree(string? lang, string? country = null)
    {
        if (string.IsNullOrWhiteSpace(lang))
        {
            return null;
        }

        string language = lang.Trim().ToLowerInvariant();
        string? ctry = string.IsNullOrWhiteSpace(country) || country.Equals("none", StringComparison.OrdinalIgnoreCase)
            ? null
            : country.Trim().ToLowerInvariant();
        string llccKey = ctry is null ? language : language + "_" + ctry;

        lock (gate)
        {
            if (cache.TryGetValue(llccKey, out HyphenationTree? cached))
            {
                return cached;
            }

            HyphenationTree? tree = LoadFromResource(llccKey);

            // Fall back to the bare language when a country-specific file is absent.
            if (tree is null && ctry is not null)
            {
                tree = GetHyphenationTree(language, null);
            }

            cache[llccKey] = tree;
            return tree;
        }
    }

    /// <summary>
    /// Hyphenates <paramref name="word"/> for the given language, or returns <c>null</c> when the
    /// language has no patterns or the word cannot be hyphenated under the given minimum counts.
    /// </summary>
    /// <param name="lang">the language code.</param>
    /// <param name="country">an optional country code.</param>
    /// <param name="word">the word to hyphenate.</param>
    /// <param name="remainCharCount">minimum characters before a hyphenation point.</param>
    /// <param name="pushCharCount">minimum characters after a hyphenation point.</param>
    public Hyphenation? Hyphenate(string? lang, string? country, string word, int remainCharCount,
        int pushCharCount)
    {
        HyphenationTree? tree = GetHyphenationTree(lang, country);
        return tree?.Hyphenate(word, remainCharCount, pushCharCount);
    }

    /// <summary>Hyphenates <paramref name="word"/> for the given language (no country).</summary>
    public Hyphenation? Hyphenate(string? lang, string word, int remainCharCount, int pushCharCount) =>
        Hyphenate(lang, null, word, remainCharCount, pushCharCount);

    private static HyphenationTree? LoadFromResource(string key)
    {
        Assembly asm = typeof(Hyphenator).Assembly;
        string resourceName = ResourcePrefix + key + ".xml";
        using Stream? stream = asm.GetManifestResourceStream(resourceName);
        if (stream is null)
        {
            return null;
        }

        var tree = new HyphenationTree();
        var settings = new XmlReaderSettings
        {
            DtdProcessing = DtdProcessing.Ignore,
            IgnoreComments = true,
            IgnoreProcessingInstructions = true,
        };
        using XmlReader reader = XmlReader.Create(stream, settings);
        tree.LoadPatterns(reader);
        return tree;
    }
}
